sqlair.go

```diff
@@ -0,0 +1,230 @@
+package sqlair
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"reflect"
+
+	"github.com/canonical/sqlair/internal/expr"
+)
+
+// Statement represents a SQL statemnt with valid SQLair expressions.
+// It is ready to be run on a SQLair DB.
+type Statement struct {
+	pe *expr.PreparedExpr
+}
+
+// Prepare expands the types mentioned in the SQLair expressions and checks
+// the SQLair parts of the query are well formed.
+// typeInstantiations must contain an instance of every type mentioned in the
+// SQLair expressions of the query. These are used only for type information.
+func Prepare(query string, typeInstantiations ...any) (*Statement, error) {
+	parser := expr.NewParser()
+	parsedExpr, err := parser.Parse(query)
+	if err != nil {
+		return nil, err
+	}
+	preparedExpr, err := parsedExpr.Prepare(typeInstantiations...)
+	if err != nil {
+		return nil, err
+	}
+	return &Statement{pe: preparedExpr}, nil
+}
+
+// MustPrepare is the same as prepare except that it panics on error.
+func MustPrepare(query string, typeInstantiations ...any) *Statement {
+	s, err := Prepare(query, typeInstantiations...)
+	if err != nil {
+		panic(err)
+	}
+	return s
+}
+
+type DB struct {
+	db *sql.DB
+}
+
+func NewDB(db *sql.DB) *DB {
+	return &DB{db: db}
+}
+
+// sqlDB returns the underlying database object.
+func (db *DB) sqlDB() *sql.DB {
+	return db.db
+}
+
+// Exec executes an SQLair Statement without returning any rows.
+// Exec uses ExecContext with context.Background internally.
+func (db *DB) Exec(s *Statement, inputStructs ...any) (sql.Result, error) {
+	return db.ExecContext(s, context.Background())
+}
+
+// ExecContext executes an SQLair Statement without returning any rows.
+func (db *DB) ExecContext(s *Statement, ctx context.Context, inputStructs ...any) (sql.Result, error) {
+	ce, err := s.pe.Complete(inputStructs...)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := db.db.ExecContext(ctx, ce.CompletedSQL(), ce.CompletedArgs()...)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// Query takes a prepared SQLair Statement and returns a Query object for
+// iterating over the results.
+// Query uses QueryContext with context.Background internally.
+func (db *DB) Query(s *Statement, inputStructs ...any) (*Query, error) {
+	return db.QueryContext(s, context.Background(), inputStructs...)
```

Comment:
Again per the conversation, let's avoid returning an error here so that this kind of logic is possible:
```
err := db.Query(...).One(&result)
```
That tends to be nice to use, and it's going to be easy as we alredy have logic around q.err.

Reply:
Changed.